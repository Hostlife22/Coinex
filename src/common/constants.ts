export const BASE_URL = 'https://project-coinex.herokuapp.com/';

export const BASE_URL_CRYPTO = 'https://api.coincap.io/';
export const CRYPTOCURRENCY_URL = 'wss://ws.coincap.io/prices?assets=ALL';

export const LOCALSTORAGE_KEY_USER = 'user';
export const LOCALSTORAGE_KEY_TOKENS = 'tokens';
export const LOCALSTORAGE_KEY_ID = 'newAccount';
export const PAGE_SIZE = 10;
export const PAGE_SIZE_SM = 3;
export const DOTS = '...';

export const MESSAGES = Object.freeze({
  signUp: {
    success: 'User Register Successfully',
    error: 'Wrong registration data',
  },
  signIn: {
    success: 'User Login Successfully',
    error: 'Invalid email or password',
  },
  validation: {
    name: 'Field must be at least 2 characters',
    email: 'Please enter a valid email address',
    password: 'Password must be at least 8 characters',
    required: {
      name: 'Field is required',
      email: 'Email is required',
      password: 'Password is required',
    },
  },
});

export const API = Object.freeze({
  auth: {
    getUrl: () => 'signin',
  },
  user: {
    getUrl: (id?: string) => `users${id ? `/${id}` : ''}`,
  },
  statistics: {
    getUrl: (id: string) => `users/${id}/statistics`,
  },
});
