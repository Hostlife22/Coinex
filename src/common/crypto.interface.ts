export interface ICoin {
  changePercent24Hr: string;
  explorer: string;
  id: string;
  marketCapUsd: string;
  maxSupply: string | null;
  name: string;
  priceUsd: string;
  rank: string;
  supply: string;
  symbol: string;
  volumeUsd24Hr: string;
  vwap24Hr: string;
}

export interface ICryptoHistory {
  priceUsd: string;
  time: number;
}

export interface IGetCryptos {
  assets: ICoin[];
}
export interface IGetCrypto {
  asset: ICoin;
}

export interface IGetCryptoHistory {
  assetHistory: ICryptoHistory[];
}
