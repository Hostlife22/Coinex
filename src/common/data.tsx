import { FaChartArea, FaHome, FaWallet } from 'react-icons/fa';
import icon from '../assets/user.png';

export const menuList = [
  {
    id: 1,
    title: 'Dashboard',
    route: '/',
    icon: <FaHome />,
  },
  {
    id: 2,
    title: 'Revenue',
    route: '/revenue',
    icon: <FaChartArea />,
  },
  {
    id: 3,
    title: 'Wallet',
    route: '/wallet',
    icon: <FaWallet />,
  },
];

export const todayExpenses = [
  {
    id: 1,
    expense: 'Monero',
    time: '5:12 pm',
    location: 'Added cryptocurrency',
    price: 326.8,
    icon: icon,
    iconBackgroundColor: '#32a7e2',
  },
  {
    id: 2,
    expense: 'TRON',
    time: '3:18 pm',
    location: 'Added cryptocurrency',
    price: 15.0,
    icon: icon,
    iconBackgroundColor: '#B548C6',
  },
  {
    id: 3,
    expense: 'Litecoin',
    time: '2:22 pm',
    location: 'Added cryptocurrency',
    price: 185.75,
    icon: icon,
    iconBackgroundColor: '#FF8700',
  },
];

export const days = [
  {
    id: 2,
    day: 7,
  },
  {
    id: 3,
    day: 14,
  },
  {
    id: 4,
    day: 30,
  },
  {
    id: 5,
    day: 60,
  },
  {
    id: 6,
    day: 120,
  },
];

export const assets = [
  {
    changePercent24Hr: '-0.9531459061618576',
    id: 'bitcoin',
    marketCapUsd: '384272416809.1633576869052544',
    maxSupply: '21000000.0000000000000000',
    name: 'Bitcoin',
    priceUsd: '20078.6670407758580608',
    rank: '1',
    supply: '19138343.0000000000000000',
    symbol: 'BTC',
    volumeUsd24Hr: '11767298311.0331411352934547',
    vwap24Hr: '20047.9813398833409349',
  },
  {
    changePercent24Hr: '-0.6092049047470719',
    id: 'ethereum',
    marketCapUsd: '192867950347.8683166675220207',
    maxSupply: null,
    name: 'Ethereum',
    priceUsd: '1578.3522587206973047',
    rank: '2',
    supply: '122195757.8115000000000000',
    symbol: 'ETH',
    volumeUsd24Hr: '7314776175.4023835649625479',
    vwap24Hr: '1563.9320035521647996',
  },
  {
    changePercent24Hr: '0.0356120675688941',
    id: 'tether',
    marketCapUsd: '67591171110.0304165951707862',
    maxSupply: null,
    name: 'Tether',
    priceUsd: '1.0006035048613313',
    rank: '3',
    supply: '67550404112.7634700000000000',
    symbol: 'USDT',
    volumeUsd24Hr: '18737250605.4503713812882649',
    vwap24Hr: '1.0003927680924454',
  },
];
