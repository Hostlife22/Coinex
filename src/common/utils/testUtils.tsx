import { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { setupApiStore } from '../../app/api/setupApiStore';
import { userApi } from '../../app/api/userApi';
import authReducer from '../../features/auth/authSlice';

interface IProps {
  children: ReactNode;
}

interface IRouterProvider extends IProps {
  route: string;
}

export const Wrapper = ({ children }: IProps) => {
  const storeRef = setupApiStore(userApi, { auth: authReducer });
  return <Provider store={storeRef.store}>{children}</Provider>;
};

export const RouterProvider = ({ children, route }: IRouterProvider) => (
  <MemoryRouter initialEntries={[route]}>
    <Routes>
      <Route path={route} element={children} />
    </Routes>
  </MemoryRouter>
);

export const ElemProvider = ({ children, route }: IRouterProvider) => {
  return (
    <Wrapper>
      <RouterProvider route={route}>{children}</RouterProvider>
    </Wrapper>
  );
};
