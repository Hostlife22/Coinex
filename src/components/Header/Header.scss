.header {
  position: fixed;
  right: 0;
  left: 0;
  z-index: 101;
  background-color: var(--bg-black-700);
  box-shadow: 0 8px 32px rgb(0 0 0 / 25%);

  &__coines {
    display: flex;
    gap: 10px;

    span {
      color: rgb(102 243 102);
    }
  }

  &__container {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 10px 40px;
  }

  &__divider {
    position: fixed;
    right: 0;
    left: 0;
    z-index: 100;
  }
}

@media screen and (max-width: 1024px) {
  .header {
    &__container {
      padding: 10px 20px;
    }
  }
}

@media screen and (max-width: 700px) {
  .header {
    &__coines {
      display: none;
    }
  }
}
