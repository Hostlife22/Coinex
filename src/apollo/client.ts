import { ApolloClient, InMemoryCache } from '@apollo/client';
import { BASE_URL_CRYPTO } from '../common/constants';

const client = new ApolloClient({
  uri: BASE_URL_CRYPTO,
  cache: new InMemoryCache(),
});

export default client;
