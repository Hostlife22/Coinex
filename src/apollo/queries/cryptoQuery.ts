import { gql } from '@apollo/client';

const GET_CRYPTOS = gql`
  query getCryptos {
    assets {
      id
      rank
      symbol
      name
      supply
      maxSupply
      marketCapUsd
      volumeUsd24Hr
      priceUsd
      changePercent24Hr
      vwap24Hr
    }
  }
`;

const GET_CRYPTO = gql`
  query getCrypto($id: ID!) {
    asset(id: $id) {
      id
      rank
      symbol
      name
      supply
      maxSupply
      marketCapUsd
      volumeUsd24Hr
      priceUsd
      changePercent24Hr
      vwap24Hr
    }
  }
`;

const GET_CRYPTO_HISTORY = gql`
  query getCryptoHistory($id: ID!, $interval: AssetHistoryInterval!) {
    assetHistory(id: $id, interval: $interval) {
      priceUsd
      time
    }
  }
`;

export { GET_CRYPTOS, GET_CRYPTO, GET_CRYPTO_HISTORY };
