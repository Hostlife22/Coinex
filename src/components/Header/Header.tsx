import { useQuery } from '@apollo/client';
import { useMemo } from 'react';
import { FaUser } from 'react-icons/fa';
import { useNavigate } from 'react-router-dom';
import { Button, Diveder, Logo, Ptag, UserPanel } from '..';
import { GET_CRYPTOS } from '../../apollo/queries/cryptoQuery';
import { useAppDispatch } from '../../app/hooks';
import { IGetCryptos } from '../../common/crypto.interface';
import { formatAsPercent } from '../../common/helpers/formatAsPercent';
import { logout } from '../../features/auth/authSlice';
import { useAuth } from '../../hooks/useAuth';
import { IHeaderProps } from './Header.interface';
import './Header.scss';

const Header = ({ handleMenu, isOpen }: IHeaderProps) => {
  const { data } = useQuery<IGetCryptos>(GET_CRYPTOS);

  const { auth } = useAuth();
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  const handleLogout = () => {
    dispatch(logout());
  };

  const currentTableData = useMemo(() => {
    return data?.assets
      ? [...data.assets].sort((a, b) => +b.changePercent24Hr - +a.changePercent24Hr).slice(0, 3)
      : [];
  }, [data?.assets]);

  return (
    <header className="header" data-testid="header">
      <div className="header__container">
        <Logo handleMenu={handleMenu} isOpen={isOpen} className="header__logo" />
        <div className="header__coines" data-testid="header-conies">
          {currentTableData.map((coin) => (
            <div key={coin.id} className="header__coin">
              <Ptag size="s">
                {coin.name} - <span>{formatAsPercent(+coin.changePercent24Hr)}</span>
              </Ptag>
            </div>
          ))}
        </div>
        {auth ? (
          <UserPanel className="header__user" isLogin={auth} logOut={handleLogout} />
        ) : (
          <Button appearance="primary" onClick={() => navigate('/auth')}>
            <FaUser /> LogIn
          </Button>
        )}
      </div>
      <Diveder className="header__divider" />
    </header>
  );
};

export default Header;
