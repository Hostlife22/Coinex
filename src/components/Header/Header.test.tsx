import { MockedProvider } from '@apollo/client/testing';
import { render, screen } from '@testing-library/react';
import MatchMediaMock from 'jest-matchmedia-mock';
import { GET_CRYPTOS } from '../../apollo/queries/cryptoQuery';
import { assets } from '../../common/data';
import { ElemProvider } from '../../common/utils/testUtils';
import Header from './Header';

let matchMedia: MatchMediaMock;
const mocks = [
  {
    request: {
      query: GET_CRYPTOS,
    },
    result: {
      data: {
        assets,
      },
    },
  },
];

const handleMenu = jest.fn();

describe('header component', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  it('renders correctly', () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ElemProvider route="/">
          <Header isOpen={true} handleMenu={handleMenu} />
        </ElemProvider>
      </MockedProvider>,
    );

    expect(screen.getByTestId('header')).toBeTruthy();
    expect(screen.getByTestId('header-conies')).toBeTruthy();
    expect(screen.getByText(/login/i)).toBeInTheDocument();
  });

  it('header snapshot', () => {
    const view = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ElemProvider route="/">
          <Header isOpen={true} handleMenu={handleMenu} />
        </ElemProvider>
      </MockedProvider>,
    );

    expect(view).toMatchSnapshot();
  });
});
