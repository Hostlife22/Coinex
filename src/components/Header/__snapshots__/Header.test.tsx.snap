// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header component header snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header"
        data-testid="header"
      >
        <div
          class="header__container"
        >
          <div
            class="header__logo logo logo_close"
            data-testid="logo"
          >
            <div
              class="logo__container"
            >
              <span
                class="logo__image"
              >
                <img
                  alt="logo"
                  src="logo.svg"
                />
              </span>
              <div
                class="logo__text "
              >
                <p
                  class="logo__name"
                  data-testid="logo-title"
                >
                  oineX
                </p>
              </div>
            </div>
            <svg
              class="logo__btn-toggle"
              data-testid="toggle"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
              />
            </svg>
          </div>
          <div
            class="header__coines"
            data-testid="header-conies"
          />
          <button
            class="btn primary"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
              />
            </svg>
             LogIn
          </button>
        </div>
        <hr
          class="header__divider diveder"
        />
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="header"
      data-testid="header"
    >
      <div
        class="header__container"
      >
        <div
          class="header__logo logo logo_close"
          data-testid="logo"
        >
          <div
            class="logo__container"
          >
            <span
              class="logo__image"
            >
              <img
                alt="logo"
                src="logo.svg"
              />
            </span>
            <div
              class="logo__text "
            >
              <p
                class="logo__name"
                data-testid="logo-title"
              >
                oineX
              </p>
            </div>
          </div>
          <svg
            class="logo__btn-toggle"
            data-testid="toggle"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
            />
          </svg>
        </div>
        <div
          class="header__coines"
          data-testid="header-conies"
        />
        <button
          class="btn primary"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
           LogIn
        </button>
      </div>
      <hr
        class="header__divider diveder"
      />
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
